// session-mongoose/index.js
// Generated by CoffeeScript 1.4.0
(function() {
  var Mongoose, Schema, SessionSchema, defaultCallback, key, mongoose, value,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  Mongoose = require('mongoose');

  mongoose = new Mongoose.Mongoose();

  for (key in Mongoose) {
    value = Mongoose[key];
    if (!(mongoose[key] != null) && Mongoose.hasOwnProperty(key)) {
      mongoose[key] = value;
    }
  }

  Schema = mongoose.Schema;

  SessionSchema = new Schema({
    sid: {
      type: String,
      required: true,
      unique: true
    },
    data: {
      type: Schema.Types.Mixed,
      required: true
    },
    expires: {
      type: Date,
      index: true
    }
  });

  defaultCallback = function(err) {};

  module.exports = function(connect) {
    var SessionStore;
    return SessionStore = (function(_super) {

      __extends(SessionStore, _super);

      function SessionStore(options) {
        var connection, _base, _base1, _ref, _ref1,
          _this = this;
        this.options = options != null ? options : {};
        if ((_ref = (_base = this.options).url) == null) {
          _base.url = "mongodb://localhost/sessions";
        }
        if ((_ref1 = (_base1 = this.options).interval) == null) {
          _base1.interval = 60000;
        }
        if (this.options.connection) {
          connection = this.options.connection;
        } else {
          if (mongoose.connection.readyState === 0) {
            connection = mongoose.connect(this.options.url);
          } else {
            connection = mongoose.connection;
          }
        }
        try {
          this.model = connection.model('Session');
        } catch (err) {
          this.model = connection.model('Session', SessionSchema);
        }
        setInterval(function() {
          return _this.model.remove({
            expires: {
              '$lte': new Date()
            }
          }, defaultCallback);
        }, this.options.interval);
      }

      SessionStore.prototype.get = function(sid, cb) {
        if (cb == null) {
          cb = defaultCallback;
        }
        return this.model.findOne({
          sid: sid
        }, function(err, session) {
          var data;
          if (err || !session) {
            return cb(err);
          } else {
            data = session.data;
            try {
              if (typeof data === 'string') {
                data = JSON.parse(data);
              }
              return cb(null, data);
            } catch (err) {
              return cb(err);
            }
          }
        });
      };

      SessionStore.prototype.set = function(sid, data, cb) {
        var cookie, expires, session;
        if (cb == null) {
          cb = defaultCallback;
        }
        if (!data) {
          return this.destroy(sid, cb);
        } else {
          try {
            if (cookie = data.cookie) {
              if (cookie.expires) {
                expires = cookie.expires;
              } else if (cookie.maxAge) {
                expires = new Date(Date.now() + cookie.maxAge);
              }
            }
            if (expires == null) {
              expires = null;
            }
            session = {
              sid: sid,
              data: data,
              expires: expires
            };
            return this.model.update({
              sid: sid
            }, session, {
              upsert: true
            }, cb);
          } catch (err) {
            return cb(err);
          }
        }
      };

      SessionStore.prototype.destroy = function(sid, cb) {
        if (cb == null) {
          cb = defaultCallback;
        }
        return this.model.remove({
          sid: sid
        }, cb);
      };

      SessionStore.prototype.all = function(cb) {
        if (cb == null) {
          cb = defaultCallback;
        }
        return this.model.find({}, 'sid expires', function(err, sessions) {
          var now, session;
          if (err || !sessions) {
            return cb(err);
          } else {
            now = Date.now();
            sessions = sessions.filter(function(session) {
              if (!session.expires || session.expires.getTime() > now) {
                return true;
              }
            });
            return cb(null, (function() {
              var _i, _len, _results;
              _results = [];
              for (_i = 0, _len = sessions.length; _i < _len; _i++) {
                session = sessions[_i];
                _results.push(session.sid);
              }
              return _results;
            })());
          }
        });
      };

      SessionStore.prototype.clear = function(cb) {
        if (cb == null) {
          cb = defaultCallback;
        }
        return this.model.collection.drop(cb);
      };

      SessionStore.prototype.length = function(cb) {
        if (cb == null) {
          cb = defaultCallback;
        }
        return this.model.count({}, cb);
      };

      return SessionStore;

    })(connect.session.Store);
  };

}).call(this);
